### DISCUSSION TITLE
Replication policy

### DISCUSSION THREAD

**Discussion by harshbhadoria011** [seaweedfs/seaweedfs#1897] [Q&A]
Hi,

We were looking for Highly available File storage system and found SeaweedFS a potential tool for our use case. 
I went through the documentation but confused regarding replication of files in volume server.
My configuration is-
Server-1: Master Volume
Server-2: Volume
Server-3: Volume

My Question is-
Suppose I'm uploading a file and at that particular time my server-2 was down, when It will come online, does all the missing data automatically gets replicated to this server or do I have to sync data from other volumes manually?

**Comment by chrislusf** [comment]
When one volume server is down, the volumes with missing replicas will become read only. The writes will be applied to writable volumes on available volume servers.

**Reply by harshbhadoria011** [reply]
What about missing files, how those files going to be replicated?
Suppose I'm writing some files and at that time one of my server was down, but when this server comes back online, how the missing files will be replicated, automatic or manually?

Before SeaweedFS I was working on HDFS, if one of the server was down during write operation, when it comes back then HDFS automatically detect that particular file/block need to be replicated.

**Reply by kmlebedev** [reply]
> What about missing files, how those files going to be replicated?
> Suppose I'm writing some files and at that time one of my server was down, but when this server comes back online, how the missing files will be replicated, automatic or manually?

Automatically and semi-automatically through a fixed scheduled job.
Two different mechanisms
1.  For each bucket, several volum files are created, which, as a rule, are located on different servers
And if one server leaves, then the volume files with one replica are transferred to read-only mode, but since there are many such folio files, there are always available for writing. 
2. If the server is absent for a long time, then the replication fix increases the number of replicas to the required

**Reply by harshbhadoria011** [reply]
As written in docs that seaweedFS can also store large file by dividing file into chunks, what is the max file size it can store without dividing into chunks.?
For ex, in our use case, file size is upto 700MB, whether seaweedFS divide this file into chunk or just keep it as one block?

**Reply by chrislusf** [reply]
See
```
weed filer -h
  ...
  -maxMB int
    	split files larger than the limit (default 32)
```

**Reply by harshbhadoria011** [reply]
While uploading file to the server, does SeaweedFS compress the file?

**Reply by chrislusf** [reply]
yes

**Reply by harshbhadoria011** [reply]
I was trying to upload a file, I did and saw that file content is being stored at one of the volumes. Volumes are files right?  which contains the data. If my uploaded file content is being stored at volume then what is the need of compressing the file?
Can you help me to understand better.

**Reply by chrislusf** [reply]
reduce required storage space

**Comment by candlerb** [comment]
> For each bucket, several volum files are created, which, as a rule, are located on different servers
And if one server leaves, then the volume files with one replica are transferred to read-only mode, but since there are many such folio files, there are always available for writing.

To add to this: my understanding, from docs and testing, is that the upload will fail entirely unless the required number of replicas are available and successfully written to *at the time of the upload*.  That is: a write only succeeds when a volume exists which meets your replication requirements (or such a volume can be created), *and* the upload is copied successfully to all replicas.  Otherwise the whole write fails.

Here's a simple test with just two volume servers, on :8080 and :8081

Case 1: one of the volume servers is down, and the incoming request asks the master for a volume with two-way replication (e.g. "001").

```
$ curl http://localhost:9333/dir/assign?replication=001
{"error":"Cannot grow volume group! No matching data node found! \nDefaultDataCenter:Only has 0 racks with more than 2 free data nodes, not enough for 1."}
```

Result: the request fails.  The master can't find or create a volume which matches your requirements.  Without a volumeID/fileID, the upload can't even start.

Case 2: the two servers are up initially, and the master gives you a volume, but the second node goes down before you complete the upload.

```
$ curl http://localhost:9333/dir/assign?replication=001
{"fid":"9,17242d59e7","url":"10.12.253.4:8080","publicUrl":"10.12.253.4:8080","count":1}

<< now shutdown the second volume server, which is actually 10.12.253.4:8081 in my test setup >>

$ curl -F file=foo 10.12.253.4:8080/9,17242d59e7
{"size":3,"error":"replicating opetations [1] is less than volume 9 replication copy count [2]","eTag":"acbd18db4cc2f85cedef654fccc4a4d8"}
```

The upload failed because the replication couldn't be completed, and no data is stored:

```
$ curl -v 10.12.253.4:8080/9,17242d59e7
...
< HTTP/1.1 404 Not Found
...
```

However, after bringing the second volume server back on line, if you retry the upload it succeeds:

```
$ curl -F file=foo 10.12.253.4:8080/9,17242d59e7
{"size":3,"eTag":"acbd18db4cc2f85cedef654fccc4a4d8"}
```

**Reply by chrislusf** [reply]
@candlerb your observation is accurate. If case 2 happens, the client should ask for a new file id from the master and retry.
